[iOS] Typing is extremely slow when editing very long sentences, with out-of-process keyboard

whsieh opened the pull request:
#### 299bc278524d3ebfb4492df41811f8ed721313a1
<pre>
[iOS] Typing is extremely slow when editing very long sentences, with out-of-process keyboard
<a href="https://bugs.webkit.org/show_bug.cgi?id=256762">https://bugs.webkit.org/show_bug.cgi?id=256762</a>
rdar://106950483

Reviewed by Myles Maxfield, Alan Baradlay and Tim Horton.

When out-of-process keyboard is enabled, UIKit requests document editing contexts upon every
selection change, and asks for three sentences-worth of context before and after the selection. For
extremely long sentences (e.g. 1000+ characters), each document context request can take hundreds of
milliseconds to process, causing the web process to noticeably hang during each keystroke. The vast
majority of this time is spent computing individual rects corresponding to each character in the
editing context range; the way we currently do this is by:

1.  Using `CharacterIterator` to construct a single-character `SimpleRange` for each character in
    the editing context range.
2.  Using `RenderObject::absoluteTextRects` on each character range, and uniting the results.
3.  Mapping each rect back into root view coordinate space.

The performance issues arise in step (2), since `RenderObject::absoluteTextRects` for a range in a
text run works by using text layout helpers to advance from the start of the run to the beginning
of the requested range, advancing to the end of the requested range, and using the difference in
width to compute the final rect. When called over *all* individual character ranges in a given text
run of length `N`, this means that the number of times we advance through the text run is `O(N^2)`.
As such, the current performance characteristics of step (2) above are `O(MN^2)`, where `M` is the
number of text runs in the context and `N` represents (on average) the number of characters per run.

To fix this, we improve the `O(N^2)` algorithm for computing character rects to just `O(N)`, by
introducing and adopting new helper methods to ask for a list of each individual character rect in
a given text run. In my testing on an iPad Pro 3rd generation with ~3000 characters worth of context
before and after the selection, this speeds document editing request time up by a factor of roughly
25x, which is enough to make typing and text interactions feel instantly responsive.

Though these changes shouldn&apos;t change behavior (since it&apos;s only a performance optimization), the
patch adds 4 new API tests to help exercise some additional corner cases that aren&apos;t already covered
by existing tests: `DocumentEditingContext.CharacterRectConsistency*`. See below for more details.

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::characterSelectionRectsForText const):

Add a new helper method to return a list of character rects, in the given range in a text run. This
works similarly to the existing `adjustSelectionRectForText` method, except that it uses a single
complex text controller to advance from the start to the end, and emits a character rect every time
it advances.

* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::absoluteTextQuads):

Refactor this to pass the whole `OptionSet` of behaviors down to `absoluteQuadsForRange`.

(WebCore::absoluteRectsForRangeInText):
* Source/WebCore/rendering/RenderObject.h:

Add a new `BoundingRectBehavior` type: `ComputeIndividualCharacterRects`, which indicates that the
client wants a rect representing each individual text character.

* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::absoluteRectsForRange const):

Refactor this to pass in an `OptionSet`.

(WebCore::characterRects const):

Add a new helper function to return a list of character rects, in the given range.

(WebCore::RenderText::absoluteQuadsForRange const):

Make this take an `OptionSet` of behaviors instead of individual boolean flags. I opted for this
approach over adding a third `bool` parameter to this function, since it would be cleaner (and make
this more extensible moving forward).

* Source/WebCore/rendering/RenderText.h:
(WebCore::RenderText::absoluteQuadsForRange):

Honor `ComputeIndividualCharacterRects` here by using the per-character helpers described above to
compute a list of character rects for each character in the given range.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Switch from `CharacterIterator` to `TextIterator`, and ask for rects for each text run found by the
text iterator using the new `ComputeIndividualCharacterRects` option.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
(-[NSString composedCharacterRanges]):
(-[UIWKDocumentContext boundingRectForCharacterRange:]):
(-[TestWKWebView firstSelectionRect]):
(-[TestWKWebView waitForFirstSelectionRectToChange:]):

Add several API test helper methods.

(checkThatAllCharacterRectsAreConsistentWithSelectionRects):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/editable-body-mixed-text.html: Added.

Add a few new test cases to verify that the character rects received via document editing context
are consistent with actual selection rects. To do this, we first use the document context API to
request character rects after focusing the test page; we then select every grapheme cluster in the
page, and verify that the final selection rect (as presented by UIKit) matches what we got in the
beginning, via the context&apos;s character rect info.

This approach keeps the tests robust over time against platform changes (as opposed to more naive
approaches like checking against hard-coded values). We also exercise various text layout scenarios
that aren&apos;t fully covered by existing tests:

- Emojis and other glyphs that span multiple codepoints
- Bidirectional text
- Left-to-right vs. right-to-left text on the body
- Horizontal vs. vertical writing mode on the body

Canonical link: <a href="https://commits.webkit.org/264086@main">https://commits.webkit.org/264086@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fec44b37a6d3229a3ece0c17729679790e6699b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8241 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6931 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6649 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7945 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6814 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9826 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6768 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7292 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8328 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4758 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6014 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13847 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6561 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6094 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8781 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6584 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5394 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5980 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1571 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10152 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6353 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
```diff
@@ -43,6 +43,29 @@ TextBoxIterator TextBox::nextTextBox() const
+        return snappedSelectionRect(characterRect, logicalRightIgnoringInlineDirection(), lineSelectionRect.y(), lineSelectionRect.height(), isHorizontal());
+    });
+    return result;
+}
```
Is there a particular reason for having this on InlineIterator::TextBox (other than that similar functions are already on TextBox)?

whsieh replied:
Indeed — the only reason is that other methods that behave similarly are already there — but since it's only called in one place, perhaps it would be cleaner to turn this into a static helper function near the call site? I'll give that a try.

alanbaradlay replied:
ok, I was planning on moving those methods out of InlineIterator::TextBox to keep this interface clean (and simple).  -they don't really belong in there.

hortont424 commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -3521,6 +3523,7 @@
 		F4CD74C820FDB49600DE3794 /* TestURLSchemeHandler.mm */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.objcpp; path = TestURLSchemeHandler.mm; sourceTree = "<group>"; };
 		F4CDAB3322489FE10057A2D9 /* UserInterfaceSwizzler.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = UserInterfaceSwizzler.h; sourceTree = "<group>"; };
 		F4CDF3D127E97C7E00191928 /* SpellCheckerDocumentTag.mm */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.objcpp; path = SpellCheckerDocumentTag.mm; sourceTree = "<group>"; };
+		F4CEF1562A102F4E00A370BE /* editable-body-mixed-text.html */ = {isa = PBXFileReference; lastKnownFileType = text.html; name = "editable-body-mixed-text.html"; path = "Tests/WebKitCocoa/editable-body-mixed-text.html"; sourceTree = SOURCE_ROOT; };
```
Somehow this file got dragged in as root-relative instead of group-relative, I think you should be able to fix it in the sidebar.

whsieh replied:
Good catch!

Edit: also moved into the `WebKit Cocoa` -> `Resources` group, for consistency with most of the other API test bundle resources.

litherum commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -400,6 +400,35 @@ bool FontCascade::fastAverageCharWidthIfAvailable(float& width) const
+        auto characterRect = selectionRect;
+        auto beforeWidth = controller.runWidthSoFar();
+
+        controller.advance(current);
+        auto afterWidth = controller.runWidthSoFar();
```
This is the right way to do it

litherum reviewed (approved):
This looks right.

whsieh committed:
[iOS] Typing is extremely slow when editing very long sentences, with out-of-process keyboard
https://bugs.webkit.org/show_bug.cgi?id=256762
rdar://106950483

Reviewed by Myles Maxfield, Alan Baradlay and Tim Horton.

When out-of-process keyboard is enabled, UIKit requests document editing contexts upon every
selection change, and asks for three sentences-worth of context before and after the selection. For
extremely long sentences (e.g. 1000+ characters), each document context request can take hundreds of
milliseconds to process, causing the web process to noticeably hang during each keystroke. The vast
majority of this time is spent computing individual rects corresponding to each character in the
editing context range; the way we currently do this is by:

1.  Using `CharacterIterator` to construct a single-character `SimpleRange` for each character in
    the editing context range.
2.  Using `RenderObject::absoluteTextRects` on each character range, and uniting the results.
3.  Mapping each rect back into root view coordinate space.

The performance issues arise in step (2), since `RenderObject::absoluteTextRects` for a range in a
text run works by using text layout helpers to advance from the start of the run to the beginning
of the requested range, advancing to the end of the requested range, and using the difference in
width to compute the final rect. When called over *all* individual character ranges in a given text
run of length `N`, this means that the number of times we advance through the text run is `O(N^2)`.
As such, the current performance characteristics of step (2) above are `O(MN^2)`, where `M` is the
number of text runs in the context and `N` represents (on average) the number of characters per run.

To fix this, we improve the `O(N^2)` algorithm for computing character rects to just `O(N)`, by
introducing and adopting new helper methods to ask for a list of each individual character rect in
a given text run. In my testing on an iPad Pro 3rd generation with ~3000 characters worth of context
before and after the selection, this speeds document editing request time up by a factor of roughly
25x, which is enough to make typing and text interactions feel instantly responsive.

Though these changes shouldn't change behavior (since it's only a performance optimization), the
patch adds 4 new API tests to help exercise some additional corner cases that aren't already covered
by existing tests: `DocumentEditingContext.CharacterRectConsistency*`. See below for more details.

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::characterSelectionRectsForText const):

Add a new helper method to return a list of character rects, in the given range in a text run. This
works similarly to the existing `adjustSelectionRectForText` method, except that it uses a single
complex text controller to advance from the start to the end, and emits a character rect every time
it advances.

* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::absoluteTextQuads):

Refactor this to pass the whole `OptionSet` of behaviors down to `absoluteQuadsForRange`.

(WebCore::absoluteRectsForRangeInText):
* Source/WebCore/rendering/RenderObject.h:

Add a new `BoundingRectBehavior` type: `ComputeIndividualCharacterRects`, which indicates that the
client wants a rect representing each individual text character.

* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::absoluteRectsForRange const):

Refactor this to pass in an `OptionSet`.

(WebCore::characterRects const):

Add a new helper function to return a list of character rects, in the given range.

(WebCore::RenderText::absoluteQuadsForRange const):

Make this take an `OptionSet` of behaviors instead of individual boolean flags. I opted for this
approach over adding a third `bool` parameter to this function, since it would be cleaner (and make
this more extensible moving forward).

* Source/WebCore/rendering/RenderText.h:
(WebCore::RenderText::absoluteQuadsForRange):

Honor `ComputeIndividualCharacterRects` here by using the per-character helpers described above to
compute a list of character rects for each character in the given range.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Switch from `CharacterIterator` to `TextIterator`, and ask for rects for each text run found by the
text iterator using the new `ComputeIndividualCharacterRects` option.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
(-[NSString composedCharacterRanges]):
(-[UIWKDocumentContext boundingRectForCharacterRange:]):
(-[TestWKWebView firstSelectionRect]):
(-[TestWKWebView waitForFirstSelectionRectToChange:]):

Add several API test helper methods.

(checkThatAllCharacterRectsAreConsistentWithSelectionRects):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/editable-body-mixed-text.html: Added.

Add a few new test cases to verify that the character rects received via document editing context
are consistent with actual selection rects. To do this, we first use the document context API to
request character rects after focusing the test page; we then select every grapheme cluster in the
page, and verify that the final selection rect (as presented by UIKit) matches what we got in the
beginning, via the context's character rect info.

This approach keeps the tests robust over time against platform changes (as opposed to more naive
approaches like checking against hard-coded values). We also exercise various text layout scenarios
that aren't fully covered by existing tests:

- Emojis and other glyphs that span multiple codepoints
- Bidirectional text
- Left-to-right vs. right-to-left text on the body
- Horizontal vs. vertical writing mode on the body

Canonical link: https://commits.webkit.org/264086@main
